service/src/main/java/com/commercetools/pspadapter/payone/transaction/BaseDefaultTransactionExecutor.java

```diff
@@ -0,0 +1,169 @@
+     * Create transaction with respective preauthorization/authorization (authorization/charge) type.
+     *
+     * @param paymentWithCartLike payment/cart which include to the request
+     * @return {@link BaseRequest} implementation with respective payment and transaction type
+     */
+    @Nonnull
+    abstract protected BaseRequest createRequest(PaymentWithCartLike paymentWithCartLike);
+
+    /**
+     * @return implementation specific logger
+     */
+    @Nonnull
+    abstract protected Logger getClassLogger();
+
+    @Override
+    public boolean wasExecuted(PaymentWithCartLike paymentWithCartLike, Transaction transaction) {
+        if (getCustomFieldsOfType(paymentWithCartLike,
+                CustomTypeBuilder.PAYONE_INTERACTION_RESPONSE,
+                CustomTypeBuilder.PAYONE_INTERACTION_REDIRECT)
+                .noneMatch(fields -> transaction.getId().equals(fields.getFieldAsString(CustomFieldKeys.TRANSACTION_ID_FIELD)))) {
+
+            return getCustomFieldsOfType(paymentWithCartLike, CustomTypeBuilder.PAYONE_INTERACTION_NOTIFICATION)
+                    //sequenceNumber field is mandatory -> can't be null
+                    .anyMatch(fields -> fields.getFieldAsString(CustomFieldKeys.SEQUENCE_NUMBER_FIELD).equals(transaction.getInteractionId()));
+        } else {
+            return true;
+        }
+    }
+
+    @Override
+    protected Optional<CustomFields> findLastExecutionAttempt(PaymentWithCartLike paymentWithCartLike, Transaction transaction) {
+        return getCustomFieldsOfType(paymentWithCartLike, CustomTypeBuilder.PAYONE_INTERACTION_REQUEST)
+                .filter(i -> transaction.getId().equals(i.getFieldAsString(CustomFieldKeys.TRANSACTION_ID_FIELD)))
+                .reduce((previous, current) -> current); // .findLast()
+    }
+
+    @Override
+    @Nonnull
+    protected PaymentWithCartLike retryLastExecutionAttempt(@Nonnull PaymentWithCartLike paymentWithCartLike,
+                                                            @Nonnull Transaction transaction,
+                                                            @Nonnull CustomFields lastExecutionAttempt) {
+        ZonedDateTime fieldAsDateTime = lastExecutionAttempt.getFieldAsDateTime(CustomFieldKeys.TIMESTAMP_FIELD);
+        if (fieldAsDateTime == null || fieldAsDateTime.isBefore(ZonedDateTime.now().minusMinutes(5))) {
+            return attemptExecution(paymentWithCartLike, transaction);
+        } else {
+            if (fieldAsDateTime.isAfter(ZonedDateTime.now().minusMinutes(1)))
+                throw new ConcurrentModificationException(String.format(
+                        "A processing of payment with ID \"%s\" started during the last 60 seconds and is likely to be finished soon, no need to retry now.",
+                        paymentWithCartLike.getPayment().getId()));
+        }
+        return paymentWithCartLike;
+    }
+
+    protected PaymentWithCartLike attemptExecution(final PaymentWithCartLike paymentWithCartLike,
```

Comment:
This method is huge! As a start, I would split the part of the try and catch to maybe a new method `AddInterfaceInteractionToPayment`:

````java
        try {
            return AddInterfaceInteractionToPayment(paymentWithCartLike, transactionId, request, updatedPayment);
        } catch (PayoneException pe) {
            getClassLogger().error("Payone request exception: ", pe);

            final AddInterfaceInteraction interfaceInteraction = AddInterfaceInteraction.ofTypeKeyAndObjects(CustomTypeBuilder.PAYONE_INTERACTION_RESPONSE,
                ImmutableMap.of(CustomFieldKeys.RESPONSE_FIELD, exceptionToResponseJsonString(pe),
                    CustomFieldKeys.TRANSACTION_ID_FIELD, transactionId,
                    CustomFieldKeys.TIMESTAMP_FIELD, ZonedDateTime.now() /* TODO */));

            final ChangeTransactionState failureTransaction = ChangeTransactionState.of(TransactionState.FAILURE, transactionId);

            return update(paymentWithCartLike, updatedPayment, ImmutableList.of(interfaceInteraction, failureTransaction));
        }
    private PaymentWithCartLike AddInterfaceInteractionToPayment(PaymentWithCartLike paymentWithCartLike,
                                                                 String transactionId, BaseRequest request,
                                                                 Payment updatedPayment) throws PayoneException {
        final Map<String, String> response = payonePostService.executePost(request);

        final String status = response.get(PayoneResponseFields.STATUS);
        if (ResponseStatus.REDIRECT.getStateCode().equals(status)) {
            final AddInterfaceInteraction interfaceInteraction = AddInterfaceInteraction.ofTypeKeyAndObjects(
                CustomTypeBuilder.PAYONE_INTERACTION_REDIRECT,
                ImmutableMap.of(CustomFieldKeys.RESPONSE_FIELD, responseToJsonString(response),
                    CustomFieldKeys.REDIRECT_URL_FIELD, response.get(PayoneResponseFields.REDIRECT_URL),
                    CustomFieldKeys.TRANSACTION_ID_FIELD, transactionId,
                    CustomFieldKeys.TIMESTAMP_FIELD, ZonedDateTime.now() /* TODO */));

            return update(paymentWithCartLike, updatedPayment,
                getRedirectUpdateActionsList(TransactionState.PENDING, updatedPayment, transactionId, response,
                    interfaceInteraction));

        } else {
            final AddInterfaceInteraction interfaceInteraction = AddInterfaceInteraction
                .ofTypeKeyAndObjects(CustomTypeBuilder.PAYONE_INTERACTION_RESPONSE,
                    ImmutableMap.of(CustomFieldKeys.RESPONSE_FIELD, responseToJsonString(response),
                        CustomFieldKeys.TRANSACTION_ID_FIELD, transactionId,
                        CustomFieldKeys.TIMESTAMP_FIELD, ZonedDateTime.now() /* TODO */));

            if (ResponseStatus.APPROVED.getStateCode().equals(status)) {
                return update(paymentWithCartLike, updatedPayment,
                    getDefaultSuccessUpdateActions(TransactionState.SUCCESS, updatedPayment, transactionId, response,
                        interfaceInteraction));
            } else if (ResponseStatus.ERROR.getStateCode().equals(status)) {
                return update(paymentWithCartLike, updatedPayment,
                    getDefaultUpdateActions(TransactionState.FAILURE, updatedPayment, transactionId, response,
                        interfaceInteraction));
            } else if (ResponseStatus.PENDING.getStateCode().equals(status)) {
                return update(paymentWithCartLike, updatedPayment,
                    getDefaultSuccessUpdateActions(TransactionState.PENDING, updatedPayment, transactionId, response,
                        interfaceInteraction));
            }
        }

        // TODO: https://github.com/commercetools/commercetools-payone-integration/issues/199
        throw new IllegalStateException("Unknown PayOne status: " + status);
    }
````

But I think even more refactoring to make it much simpler could be done here.

Reply:
not in this PR

Reply:
do we need to make issues about such "not in this PR" comments?

Reply:
https://github.com/commercetools/commercetools-payone-integration/issues/204